### DISCUSSION TITLE
Working with ADMaterial

### DISCUSSION THREAD

**Discussion by rksin8** [idaholab/moose#18482] [Q&A Modules: General] [answered]
Is there any equivalent AD function to the `coupledValueOld()` and `getMaterialPropertyOld<T>()`?

Thanks,
Ranjeet

**Comment by rksin8** [comment]
Solved this issue with help of tensor module material.

**Comment by dschwen** [comment, marked as answer]
Old state properties are always non-dual numbers. The old state of an `ADReal` property is simply of type `Real`. That is because the old state does not depend on the current solution state! So the solution is to simply use `coupledValueOld()` and `getMaterialPropertyOld<T>()`.

**Reply by rksin8** [reply]
Thank you for explaining this. 

Would you please explain when to use precomputeQpResidual() instead of computeQpResidual()?

Thank you!

**Reply by YaqiWang** [reply]
You should start a new thread for this question. I assume you are asking something like `precalculateQpResidual` in `IntegratedBC` vs `computeQpResidual`. `precalculateQpResidual` provides derived objects a function to override where they can put local evaluations independent on shape functions (`_i`) to avoid repeated operations. You can see this by checking the code in `IntegratedBC.C`.

**Reply by rksin8** [reply]
@YaqiWang Thank you for pointing out some reference codes. I will ask a new question if need further help.
